File: src/test/java/ru/kiseru/checkers/board/ManTest.java
package ru.kiseru.checkers.board;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.test.util.ReflectionTestUtils;
import ru.kiseru.checkers.exception.CannotEatException;
import ru.kiseru.checkers.exception.CannotMoveException;
import ru.kiseru.checkers.exception.MustEatException;
import ru.kiseru.checkers.utils.Color;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatExceptionOfType;
import static org.mockito.BDDMockito.any;
import static org.mockito.BDDMockito.anyInt;
import static org.mockito.BDDMockito.eq;
import static org.mockito.BDDMockito.given;
import static org.mockito.BDDMockito.mock;
import static org.mockito.BDDMockito.willCallRealMethod;

@ExtendWith(MockitoExtension.class)
class ManTest {

    private static final int SOURCE_CELL_COLUMN = 4;

    private static final int SOURCE_CELL_ROW = 4;

    @InjectMocks
    private Man underTest = new Man(Color.WHITE);

    @Mock
    private Board board;

    @Mock
    private Cell cell;

    @Mock
    private Cell destinationCell;

    @BeforeEach
    void setUp() {
        ReflectionTestUtils.setField(cell, "piece", underTest);
    }

    @Test
    void testIsAbleToMoveToWhileDestinationCellIsNotEmpty() {
        // given
        given(destinationCell.getPiece()).willReturn(mock(Piece.class));

        // when
        var actual = underTest.isAbleToMoveTo(destinationCell);

        // then
        assertThat(actual).isFalse();
    }

    @Test
    void testIsAbleToMoveToWhileEnemyIsNearby() {
        // given
        var enemy = mock(Man.class);
        ReflectionTestUtils.setField(enemy, "color", Color.BLACK);

        var cellWithEnemy = mock(Cell.class);
        given(cellWithEnemy.getPiece()).willReturn(enemy);

        given(board.getCell(anyInt(), anyInt())).willReturn(cellWithEnemy);

        given(cell.getBoard()).willReturn(board);

        // when
        var actual = underTest.isAbleToMoveTo(destinationCell);

        // then
        assertThat(actual).isFalse();
    }

    @Test
    void testIsAbleToMoveToWhileDestinationDiffIsMoreThanOne() {
        // given
        var emptyCell = mock(Cell.class);
        given(emptyCell.getPiece()).willReturn(null);

        given(board.getCell(anyInt(), anyInt())).willReturn(emptyCell);

        given(cell.getBoard()).willReturn(board);
        given(cell.diff(eq(destinationCell))).willReturn(2);

        // when
        var actual = underTest.isAbleToMoveTo(destinationCell);

        // then
        assertThat(actual).isFalse();
    }

    @Test
    void testIsAbleToMoveToWhileThereIsNoNextRow() {
        // given
        var emptyCell = mock(Cell.class);
        given(emptyCell.getPiece()).willReturn(null);

        given(board.getCell(anyInt(), anyInt())).willReturn(emptyCell);

        given(cell.getBoard()).willReturn(board);
        given(cell.diff(eq(destinationCell))).willReturn(1);
        given(cell.getRow()).willReturn(8);

        // when
        var actual = underTest.isAbleToMoveTo(destinationCell);

        // then
        assertThat(actual).isFalse();
    }

    @Test
    void testIsAbleToMoveToWhileCanMoveToDestination() {
        // given
        var emptyCell = mock(Cell.class);
        given(emptyCell.getPiece()).willReturn(null);

        given(board.getCell(eq(SOURCE_CELL_ROW + 1), eq(SOURCE_CELL_COLUMN - 1))).willReturn(destinationCell);
        given(board.getCell(eq(SOURCE_CELL_ROW + 1), eq(SOURCE_CELL_COLUMN + 1))).willReturn(destinationCell);
        given(board.getCell(eq(SOURCE_CELL_ROW - 1), eq(SOURCE_CELL_COLUMN - 1))).willReturn(emptyCell);
        given(board.getCell(eq(SOURCE_CELL_ROW - 1), eq(SOURCE_CELL_COLUMN + 1))).willReturn(emptyCell);

        given(cell.getColumn()).willReturn(SOURCE_CELL_COLUMN);
        given(cell.getRow()).willReturn(SOURCE_CELL_ROW);
        given(cell.getBoard()).willReturn(board);
        given(cell.diff(eq(destinationCell))).willReturn(1);

        // when
        var actual = underTest.isAbleToMoveTo(destinationCell);

        // then
        assertThat(actual).isTrue();
    }

    @Test
    void testIsAbleToEatToWhileCellDiffIsLessThanTwo() {
        // given
        given(cell.diff(eq(destinationCell))).willReturn(1);

        // when
        var actual = underTest.isAbleToEatTo(destinationCell);

        // then
        assertThat(actual).isFalse();
    }

    @Test
    void testIsAbleToEatToWhileCellDiffIsMoreThanTwo() {
        // given
        given(cell.diff(eq(destinationCell))).willReturn(3);

        // when
        var actual = underTest.isAbleToEatTo(destinationCell);

        // then
        assertThat(actual).isFalse();
    }

    @Test
    void testIsAbleToEatToWhileDestinationCellIsNotBusy() {
        // given
        given(cell.diff(eq(destinationCell))).willReturn(2);

        given(destinationCell.getPiece()).willReturn(mock(Piece.class));

        // when
        var actual = underTest.isAbleToEatTo(destinationCell);

        // then
        assertThat(actual).isFalse();
    }

    @Test
    void testIsAbleToEatToWhileThereIsNoSacrificePiece() {
        // given
        var cellWithSacrifice = mock(Cell.class);
        given(cellWithSacrifice.getPiece()).willReturn(null);

        given(cell.diff(eq(destinationCell))).willReturn(2);
        given(cell.getBoard()).willReturn(board);
        given(cell.between(eq(destinationCell), eq(board))).willReturn(cellWithSacrifice);

        // when
        var actual = underTest.isAbleToEatTo(destinationCell);

        // then
        assertThat(actual).isFalse();
    }

    @Test
    void testIsAbleToEatToWhileSacrificePieceHasTheSameColor() {
        // given
        var sacrificePiece = mock(Man.class);
        given(sacrificePiece.getColor()).willReturn(Color.WHITE);

        var cellWithSacrifice = mock(Cell.class);
        given(cellWithSacrifice.getPiece()).willReturn(sacrificePiece);

        given(cell.diff(eq(destinationCell))).willReturn(2);
        given(cell.getBoard()).willReturn(board);
        given(cell.between(eq(destinationCell), eq(board))).willReturn(cellWithSacrifice);

        // when
        var actual = underTest.isAbleToEatTo(destinationCell);

        // then
        assertThat(actual).isFalse();
    }

    @Test
    void testIsAbleToEatToWhileSacrificePieceHasEnemyColor() {
        // given
        var sacrificePiece = mock(Man.class);
        ReflectionTestUtils.setField(sacrificePiece, "color", Color.BLACK);

        var cellWithSacrifice = mock(Cell.class);
        given(cellWithSacrifice.getPiece()).willReturn(sacrificePiece);

        given(cell.diff(eq(destinationCell))).willReturn(2);
        given(cell.getBoard()).willReturn(board);
        given(cell.between(eq(destinationCell), eq(board))).willReturn(cellWithSacrifice);

        // when
        var actual = underTest.isAbleToEatTo(destinationCell);

        // then
        assertThat(actual).isTrue();
    }

    @Test
    void testAnalyzeAbilityOfMove() {
        // given
        var emptyCell = mock(Cell.class);

        given(cell.getBoard()).willReturn(board);
        given(cell.getColumn()).willReturn(SOURCE_CELL_COLUMN);
        given(cell.getRow()).willReturn(SOURCE_CELL_ROW);
        given(cell.diff(eq(emptyCell))).willReturn(1);

        given(board.getCell(anyInt(), anyInt())).willReturn(emptyCell);

        // when
        underTest.analyzeAbilityOfMove();

        // then
        var canMove = underTest.isCanMove();
        assertThat(canMove).isEqualTo(Boolean.TRUE);
    }

    @Test
    void testAnalyzeAbilityOfEat() {
        // given
        var emptyCell = mock(Cell.class);

        var enemyPiece = mock(Piece.class);
        ReflectionTestUtils.setField(enemyPiece, "color", Color.BLACK);

        var cellWithEnemyPiece = mock(Cell.class);
        given(cellWithEnemyPiece.getPiece()).willReturn(enemyPiece);

        given(cell.getBoard()).willReturn(board);
        given(cell.getColumn()).willReturn(SOURCE_CELL_COLUMN);
        given(cell.getRow()).willReturn(SOURCE_CELL_ROW);
        given(cell.diff(eq(emptyCell))).willReturn(2);
        given(cell.between(eq(emptyCell), eq(board))).willReturn(cellWithEnemyPiece);

        given(board.getCell(anyInt(), anyInt())).willReturn(emptyCell);

        // when
        underTest.analyzeAbilityOfEat();

        // then
        var canEat = underTest.isCanEat();
        assertThat(canEat).isEqualTo(Boolean.TRUE);
    }

    @Test
    void testMoveWhileCanEat() {
        // given
        underTest.setCanEat(true);

        // then
        assertThatExceptionOfType(MustEatException.class)
                .isThrownBy(() -> underTest.move(destinationCell));
    }

    @Test
    void testMoveWhileCannotMove() {
        // given
        underTest.setCanEat(false);
        underTest.setCanMove(false);

        // then
        assertThatExceptionOfType(CannotMoveException.class)
                .isThrownBy(() -> underTest.move(destinationCell));
    }

    @Test
    void testMoveWhileCannotMoveTo() {
        // given
        underTest.setCanEat(false);
        underTest.setCanMove(true);

        given(destinationCell.getPiece()).willReturn(mock(Piece.class));

        // then
        assertThatExceptionOfType(CannotMoveException.class)
                .isThrownBy(() -> underTest.move(destinationCell));
    }

    @Test
    void testMove() {
        // given
        underTest.setCanEat(false);
        underTest.setCanMove(true);

        var emptyCell = mock(Cell.class);

        given(board.getCell(anyInt(), anyInt())).willReturn(emptyCell);
        given(board.getCell(eq(1), eq(1))).willReturn(destinationCell);

        willCallRealMethod().given(cell).setPiece(any());
        given(cell.getBoard()).willReturn(board);
        given(cell.diff(eq(destinationCell))).willReturn(1);


        willCallRealMethod().given(destinationCell).setPiece(any());

        // when
        underTest.move(destinationCell);

        // then
        var pieceInDestinationCell = ReflectionTestUtils.getField(destinationCell, "piece");
        assertThat(pieceInDestinationCell).isEqualTo(underTest);

        var pieceInSourceCell = ReflectionTestUtils.getField(cell, "piece");
        assertThat(pieceInSourceCell).isNull();
    }

    @Test
    void testMoveWhileWhitePieceBecomeKing() {
        // given
        underTest.setCanEat(false);
        underTest.setCanMove(true);

        var emptyCell = mock(Cell.class);

        given(board.getCell(anyInt(), anyInt())).willReturn(emptyCell);
        given(board.getCell(eq(8), eq(2))).willReturn(destinationCell);

        given(cell.getBoard()).willReturn(board);
        given(cell.diff(eq(destinationCell))).willReturn(1);
        given(cell.getRow()).willReturn(7);
        given(cell.getColumn()).willReturn(1);

        given(destinationCell.getRow()).willReturn(8);

        willCallRealMethod().given(destinationCell).setPiece(any());
        willCallRealMethod().given(cell).setPiece(any());

        // when
        underTest.move(destinationCell);

        // then
        var pieceInDestinationCell = ReflectionTestUtils.getField(destinationCell, "piece");
        assertThat(pieceInDestinationCell).isNotNull();
        assertThat(pieceInDestinationCell).isInstanceOf(King.class);
        assertThat(((King) pieceInDestinationCell).getColor()).isEqualTo(Color.WHITE);

        var pieceInSourceCell = ReflectionTestUtils.getField(cell, "piece");
        assertThat(pieceInSourceCell).isNull();
    }

    @Test
    void testMoveWhileBlackPieceBecomeKing() {
        // given
        underTest.setCanEat(false);
        underTest.setCanMove(true);

        var emptyCell = mock(Cell.class);

        given(board.getCell(anyInt(), anyInt())).willReturn(emptyCell);
        given(board.getCell(eq(1), eq(1))).willReturn(destinationCell);

        given(cell.getBoard()).willReturn(board);
        given(cell.diff(eq(destinationCell))).willReturn(1);
        given(cell.getRow()).willReturn(2);
        given(cell.getColumn()).willReturn(2);

        given(destinationCell.getRow()).willReturn(1);

        willCallRealMethod().given(destinationCell).setPiece(any());
        willCallRealMethod().given(cell).setPiece(any());

        ReflectionTestUtils.setField(underTest, "color", Color.BLACK);

        // when
        underTest.move(destinationCell);

        // then
        var pieceInDestinationCell = ReflectionTestUtils.getField(destinationCell, "piece");
        assertThat(pieceInDestinationCell).isNotNull();
        assertThat(pieceInDestinationCell).isInstanceOf(King.class);
        assertThat(((King) pieceInDestinationCell).getColor()).isEqualTo(Color.BLACK);

        var pieceInSourceCell = ReflectionTestUtils.getField(cell, "piece");
        assertThat(pieceInSourceCell).isNull();
    }

    @Test
    void testEatWhileCannotEat() {
        // given
        underTest.setCanEat(false);

        // then
        assertThatExceptionOfType(CannotEatException.class)
                .isThrownBy(() -> underTest.eat(destinationCell));
    }

    @Test
    void testEatWhileCannotEatAtSomeDestination() {
        // given
        underTest.setCanEat(true);

        // then
        assertThatExceptionOfType(CannotEatException.class)
                .isThrownBy(() -> underTest.eat(destinationCell));
    }

    @Test
    void testEat() {
        // given
        var enemy = mock(Piece.class);
        ReflectionTestUtils.setField(enemy, "color", Color.BLACK);

        var targetCell = mock(Cell.class);
        given(targetCell.getPiece()).willReturn(enemy);
        willCallRealMethod().given(targetCell).setPiece(any());

        given(cell.between(eq(destinationCell), any())).willReturn(targetCell);
        given(cell.diff(eq(destinationCell))).willReturn(2);
        willCallRealMethod().given(cell).setPiece(any());

        underTest.setCanEat(true);

        willCallRealMethod().given(destinationCell).setPiece(any());

        // when
        underTest.eat(destinationCell);

        // then
        var pieceInSourceCell = ReflectionTestUtils.getField(cell, "piece");
        assertThat(pieceInSourceCell).isNull();

        var pieceInTargetCell = ReflectionTestUtils.getField(targetCell, "piece");
        assertThat(pieceInTargetCell).isNull();

        var pieceInDestinationCell = ReflectionTestUtils.getField(destinationCell, "piece");
        assertThat(pieceInDestinationCell).isEqualTo(underTest);
    }

    @Test
    void testEatWhileWhiteManShouldBecomeKing() {
        // given
        given(destinationCell.getRow()).willReturn(8);
        willCallRealMethod().given(destinationCell).setPiece(any());

        var enemy = mock(Piece.class);
        ReflectionTestUtils.setField(enemy, "color", Color.BLACK);

        var targetCell = mock(Cell.class);
        given(targetCell.getPiece()).willReturn(enemy);
        willCallRealMethod().given(targetCell).setPiece(any());

        given(cell.between(eq(destinationCell), any())).willReturn(targetCell);
        given(cell.diff(eq(destinationCell))).willReturn(2);
        willCallRealMethod().given(cell).setPiece(any());

        underTest.setCanEat(true);

        // when
        underTest.eat(destinationCell);

        // then
        var pieceInSourceCell = ReflectionTestUtils.getField(cell, "piece");
        assertThat(pieceInSourceCell).isNull();

        var pieceInTargetCell = ReflectionTestUtils.getField(targetCell, "piece");
        assertThat(pieceInTargetCell).isNull();

        var pieceInDestinationCell = ReflectionTestUtils.getField(destinationCell, "piece");
        assertThat(pieceInDestinationCell).isNotNull();
        assertThat(pieceInDestinationCell).isInstanceOf(King.class);
        assertThat(((King) pieceInDestinationCell).getColor()).isEqualTo(Color.WHITE);
    }

    @Test
    void testEatWhileBlackManShouldBecomeKing() {
        // given
        given(destinationCell.getRow()).willReturn(1);
        willCallRealMethod().given(destinationCell).setPiece(any());

        var enemy = mock(Piece.class);
        ReflectionTestUtils.setField(enemy, "color", Color.WHITE);

        var targetCell = mock(Cell.class);
        given(targetCell.getPiece()).willReturn(enemy);
        willCallRealMethod().given(targetCell).setPiece(any());

        given(cell.between(eq(destinationCell), any())).willReturn(targetCell);
        given(cell.diff(eq(destinationCell))).willReturn(2);
        willCallRealMethod().given(cell).setPiece(any());

        ReflectionTestUtils.setField(underTest, "color", Color.BLACK);
        underTest.setCanEat(true);

        // when
        underTest.eat(destinationCell);

        // then
        var pieceInSourceCell = ReflectionTestUtils.getField(cell, "piece");
        assertThat(pieceInSourceCell).isNull();

        var pieceInTargetCell = ReflectionTestUtils.getField(targetCell, "piece");
        assertThat(pieceInTargetCell).isNull();

        var pieceInDestinationCell = ReflectionTestUtils.getField(destinationCell, "piece");
        assertThat(pieceInDestinationCell).isNotNull();
        assertThat(pieceInDestinationCell).isInstanceOf(King.class);
        assertThat(((King) pieceInDestinationCell).getColor()).isEqualTo(Color.BLACK);
    }

    @Test
    void testCreating() {
        // when
        var man = new Man(Color.BLACK);

        // then
        assertThat(man.getColor()).isEqualTo(Color.BLACK);
    }

    @Test
    void testIsMan() {
        // when
        var actual = underTest.isMan();

        // then
        assertThat(actual).isTrue();
    }

    @Test
    void testIsKing() {
        // when
        var actual = underTest.isKing();

        // then
        assertThat(actual).isFalse();
    }
}

File: src/test/kotlin/ru/kiseru/checkers/board/ManTest.kt
package ru.kiseru.checkers.board

import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatExceptionOfType
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.ArgumentMatchers.anyInt
import org.mockito.BDDMockito.willCallRealMethod
import org.mockito.InjectMocks
import org.mockito.Mock
import org.mockito.junit.jupiter.MockitoExtension
import org.mockito.kotlin.any
import org.mockito.kotlin.anyOrNull
import org.mockito.kotlin.eq
import org.mockito.kotlin.given
import org.mockito.kotlin.mock
import org.springframework.test.util.ReflectionTestUtils
import ru.kiseru.checkers.exception.CannotEatException
import ru.kiseru.checkers.exception.CannotMoveException
import ru.kiseru.checkers.exception.MustEatException
import ru.kiseru.checkers.utils.Color

@ExtendWith(MockitoExtension::class)
class ManTest {

    private val sourceCellColumn = 4

    private val sourceCellRow = 4

    @InjectMocks
    private val underTest = Man(Color.WHITE)

    @Mock
    private lateinit var board: Board

    @Mock
    private lateinit var cell: Cell

    @Mock
    private lateinit var destinationCell: Cell

    @Test
    fun `isAbleToMoveTo when destination cell isn't empty`() {
        // given
        given(destinationCell.piece).willReturn(mock<Piece>())

        // when
        val actual = underTest.isAbleToMoveTo(destinationCell)

        // then
        assertThat(actual).isFalse()
    }

    @Test
    fun `isAbleToMoveTo test when enemy is nearby`() {
        // given
        val enemy = mock<Man>()
        ReflectionTestUtils.setField(enemy, "color", Color.BLACK)

        val cellWithEnemy = mock<Cell>()
        given(cellWithEnemy.piece).willReturn(enemy)

        given(board.getCell(anyInt(), anyInt())).willReturn(cellWithEnemy)

        given(cell.board).willReturn(board)

        // when
        val actual = underTest.isAbleToMoveTo(destinationCell)

        // then
        assertThat(actual).isFalse()
    }

    @Test
    fun `isAbleToMoveTo test when there is no next row`() {
        // given
        val emptyCell = mock<Cell>()

        given(board.getCell(anyInt(), anyInt())).willReturn(emptyCell)

        given(cell.board).willReturn(board)
        given(cell.diff(eq(destinationCell))).willReturn(1)
        given(cell.row).willReturn(8)

        // when
        val actual = underTest.isAbleToMoveTo(destinationCell)

        // then
        assertThat(actual).isFalse()
    }

    @Test
    fun `isAbleToMoveTo test when the player can move to destination cell`() {
        // given
        val emptyCell = mock<Cell>()

        given(board.getCell(eq(sourceCellRow + 1), eq(sourceCellColumn + 1))).willReturn(destinationCell)
        given(board.getCell(eq(sourceCellRow + 1), eq(sourceCellColumn - 1))).willReturn(destinationCell)
        given(board.getCell(eq(sourceCellRow - 1), eq(sourceCellColumn - 1))).willReturn(emptyCell)
        given(board.getCell(eq(sourceCellRow - 1), eq(sourceCellColumn + 1))).willReturn(emptyCell)

        given(cell.column).willReturn(sourceCellRow)
        given(cell.row).willReturn(sourceCellColumn)
        given(cell.board).willReturn(board)
        given(cell.diff(eq(destinationCell))).willReturn(1)

        // when
        val actual = underTest.isAbleToMoveTo(destinationCell)

        // then
        assertThat(actual).isTrue()
    }

    @Test
    fun `isAbleToEatTo test when diff is less than two`() {
        // given
        given(cell.diff(eq(destinationCell))).willReturn(1)

        // when
        val actual = underTest.isAbleToEatTo(destinationCell)

        // then
        assertThat(actual).isFalse()
    }

    @Test
    fun `isAbleToMoveTo test when diff is more than two`() {
        // given
        given(cell.diff(eq(destinationCell))).willReturn(3)

        // when
        val actual = underTest.isAbleToEatTo(destinationCell)

        // then
        assertThat(actual).isFalse()
    }

    @Test
    fun `isAbleToEatTo test when destination cell isn't busy`() {
        // given
        given(cell.diff(eq(destinationCell))).willReturn(2)

        given(destinationCell.piece).willReturn(mock<Piece>())

        // when
        val actual = underTest.isAbleToEatTo(destinationCell)

        // then
        assertThat(actual).isFalse()
    }

    @Test
    fun `isAbleToEatTo test when there are no sacrifice pieces`() {
        // given
        val cellWithSacrifice = mock<Cell>()

        given(cell.diff(eq(destinationCell))).willReturn(2)
        given(cell.between(eq(destinationCell))).willReturn(cellWithSacrifice)

        // when
        val actual = underTest.isAbleToEatTo(destinationCell)

        // then
        assertThat(actual).isFalse()
    }

    @Test
    fun `isAbleToEatTo test when sacrifice piece has the same color`() {
        // given
        val sacrificePiece = mock<Man>()
        given(sacrificePiece.color).willReturn(Color.WHITE)

        val cellWithSacrifice = mock<Cell>()
        given(cellWithSacrifice.piece).willReturn(sacrificePiece)

        given(cell.diff(eq(destinationCell))).willReturn(2)
        given(cell.between(eq(destinationCell))).willReturn(cellWithSacrifice)

        // when
        val actual = underTest.isAbleToEatTo(destinationCell)

        // then
        assertThat(actual).isFalse()
    }

    @Test
    fun `isAbleToEatTo test when sacrifice piece has enemy color`() {
        // given
        val sacrificePiece = mock<Man>()
        ReflectionTestUtils.setField(sacrificePiece, "color", Color.BLACK)

        val cellWithSacrifice = mock<Cell>()
        given(cellWithSacrifice.piece).willReturn(sacrificePiece)

        given(cell.diff(eq(destinationCell))).willReturn(2)
        given(cell.between(eq(destinationCell))).willReturn(cellWithSacrifice)

        // when
        val actual = underTest.isAbleToEatTo(destinationCell)

        // then
        assertThat(actual).isTrue()
    }

    @Test
    fun `analyzeAbilityOfMove test`() {
        // given
        val emptyCell = mock<Cell>()

        given(cell.board).willReturn(board)
        given(cell.column).willReturn(sourceCellColumn)
        given(cell.row).willReturn(sourceCellRow)
        given(cell.diff(eq(emptyCell))).willReturn(1)

        given(board.getCell(anyInt(), anyInt())).willReturn(emptyCell)

        // when
        underTest.analyzeAbilityOfMove()

        // then
        assertThat(underTest.isCanMove).isTrue()
    }

    @Test
    fun `analyzeAbilityOfEat test`() {
        // given
        val emptyCell = mock<Cell>()

        val enemyPiece = mock<Piece>()
        ReflectionTestUtils.setField(enemyPiece, "color", Color.BLACK)

        val cellWithEnemyPiece = mock<Cell>()
        given(cellWithEnemyPiece.piece).willReturn(enemyPiece)

        given(cell.board).willReturn(board)
        given(cell.column).willReturn(sourceCellColumn)
        given(cell.row).willReturn(sourceCellRow)
        given(cell.diff(eq(emptyCell))).willReturn(2)
        given(cell.between(eq(emptyCell))).willReturn(cellWithEnemyPiece)

        given(board.getCell(anyInt(), anyInt())).willReturn(emptyCell)

        // when
        underTest.analyzeAbilityOfEat()

        // then
        assertThat(underTest.isCanEat).isTrue()
    }

    @Test
    fun `move test when man can eat`() {
        // given
        underTest.isCanEat = true

        // when & then
        assertThatExceptionOfType(MustEatException::class.java)
            .isThrownBy { underTest.move(destinationCell) }
    }

    @Test
    fun `move test when man can't move`() {
        // given
        underTest.isCanEat = false
        underTest.isCanMove = false

        // then
        assertThatExceptionOfType(CannotMoveException::class.java)
            .isThrownBy { underTest.move(destinationCell) }
    }

    @Test
    fun `move test when man can't move to`() {
        // given
        underTest.isCanEat = false
        underTest.isCanMove = true

        given(destinationCell.piece).willReturn(mock<Piece>())

        // when & then
        assertThatExceptionOfType(CannotMoveException::class.java)
            .isThrownBy { underTest.move(destinationCell) }
    }

    @Test
    fun `move test`() {
        // given
        underTest.isCanEat = false
        underTest.isCanMove = true

        given(board.getCell(anyInt(), anyInt())).willReturn(mock<Cell>())
        given(board.getCell(eq(1), eq(1))).willReturn(destinationCell)

        willCallRealMethod().given(cell).piece = anyOrNull()
        given(cell.board).willReturn(board)
        given(cell.diff(eq(destinationCell))).willReturn(1)

        willCallRealMethod().given(destinationCell).piece = any<Piece>()

        // when
        underTest.move(destinationCell)

        // then
        val pieceInDestinationCell = ReflectionTestUtils.getField(destinationCell, "piece")
        assertThat(pieceInDestinationCell).isSameAs(underTest)

        val pieceInSourceCell = ReflectionTestUtils.getField(cell, "piece")
        assertThat(pieceInSourceCell).isNull()
    }

    @Test
    fun `move test when white piece becomes king`() {
        // given
        underTest.isCanEat = false
        underTest.isCanMove = true

        given(board.getCell(anyInt(), anyInt())).willReturn(mock<Cell>())
        given(board.getCell(eq(8), eq(2))).willReturn(destinationCell)

        given(cell.board).willReturn(board)
        given(cell.diff(eq(destinationCell))).willReturn(1)
        given(cell.row).willReturn(7)
        given(cell.column).willReturn(1)
        willCallRealMethod().given(cell).piece = anyOrNull()

        given(destinationCell.row).willReturn(8)
        willCallRealMethod().given(destinationCell).piece = any<Piece>()

        // when
        underTest.move(destinationCell)

        // then
        val pieceInDestinationCell = ReflectionTestUtils.getField(destinationCell, "piece")
        assertThat(pieceInDestinationCell).isNotNull()
        assertThat(pieceInDestinationCell).isInstanceOf(King::class.java)
        assertThat((pieceInDestinationCell as King).color).isEqualTo(Color.WHITE)

        val pieceInSourceCell = ReflectionTestUtils.getField(cell, "piece")
        assertThat(pieceInSourceCell).isNull()
    }

    @Test
    fun `move test when black piece becomes king`() {
        // given
        underTest.isCanEat = false
        underTest.isCanMove = true
        ReflectionTestUtils.setField(underTest, "color", Color.BLACK)

        given(board.getCell(anyInt(), anyInt())).willReturn(mock<Cell>())
        given(board.getCell(eq(1), eq(1))).willReturn(destinationCell)

        given(cell.board).willReturn(board)
        given(cell.diff(eq(destinationCell))).willReturn(1)
        given(cell.row).willReturn(2)
        given(cell.column).willReturn(2)
        willCallRealMethod().given(cell).piece = anyOrNull()

        given(destinationCell.row).willReturn(1)
        willCallRealMethod().given(destinationCell).piece = any<Piece>()

        // when
        underTest.move(destinationCell)

        // then
        val pieceInDestinationCell = ReflectionTestUtils.getField(destinationCell, "piece")
        assertThat(pieceInDestinationCell).isNotNull()
        assertThat(pieceInDestinationCell).isInstanceOf(King::class.java)
        assertThat((pieceInDestinationCell as King).color).isEqualTo(Color.BLACK)

        val pieceInSourceCell = ReflectionTestUtils.getField(cell, "piece")
        assertThat(pieceInSourceCell).isNull()
    }

    @Test
    fun `eat test when piece can't eat`() {
        // given
        underTest.isCanEat = false

        // when & then
        assertThatExceptionOfType(CannotEatException::class.java)
            .isThrownBy { underTest.eat(destinationCell) }
    }

    @Test
    fun `eat test when piece can't eat at some destination`() {
        // given
        underTest.isCanEat = true

        // then
        assertThatExceptionOfType(CannotEatException::class.java)
            .isThrownBy { underTest.eat(destinationCell) }
    }

    @Test
    fun `eat test`() {
        // given
        val enemy = mock<Piece>()
        ReflectionTestUtils.setField(enemy, "color", Color.BLACK)

        val targetCell = mock<Cell>()
        given(targetCell.piece).willReturn(enemy)
        willCallRealMethod().given(targetCell).piece = anyOrNull()

        given(cell.between(eq(destinationCell))).willReturn(targetCell)
        given(cell.diff(eq(destinationCell))).willReturn(2)
        willCallRealMethod().given(cell).piece = anyOrNull()

        underTest.isCanEat = true

        willCallRealMethod().given(destinationCell).piece = any()

        // when
        underTest.eat(destinationCell)

        // then
        val pieceInSourceCell = ReflectionTestUtils.getField(cell, "piece")
        assertThat(pieceInSourceCell).isNull()

        val pieceInTargetCell = ReflectionTestUtils.getField(targetCell, "piece")
        assertThat(pieceInTargetCell).isNull()

        val pieceInDestinationCell = ReflectionTestUtils.getField(destinationCell, "piece")
        assertThat(pieceInDestinationCell).isSameAs(underTest)
    }

    @Test
    fun `eat test when white man should become king`() {
        // given
        given(destinationCell.row).willReturn(8)
        willCallRealMethod().given(destinationCell).piece = any<Piece>()

        val enemy = mock<Piece>()
        ReflectionTestUtils.setField(enemy, "color", Color.BLACK)

        val targetCell = mock<Cell>()
        given(targetCell.piece).willReturn(enemy)
        willCallRealMethod().given(targetCell).piece = anyOrNull()

        given(cell.between(eq(destinationCell))).willReturn(targetCell)
        given(cell.diff(eq(destinationCell))).willReturn(2)
        willCallRealMethod().given(cell).piece = anyOrNull()

        underTest.isCanEat = true

        // when
        underTest.eat(destinationCell)

        // then
        val pieceInSourceCell = ReflectionTestUtils.getField(cell, "piece")
        assertThat(pieceInSourceCell).isNull()

        val pieceInTargetCell = ReflectionTestUtils.getField(targetCell, "piece")
        assertThat(pieceInTargetCell).isNull()

        val pieceInDestinationCell = ReflectionTestUtils.getField(destinationCell, "piece")
        assertThat(pieceInDestinationCell).isNotNull()
        assertThat(pieceInDestinationCell).isInstanceOf(King::class.java)
        assertThat((pieceInDestinationCell as King).color).isEqualTo(Color.WHITE)
    }

    @Test
    fun `eat test when black man should become king`() {
        // given
        given(destinationCell.row).willReturn(1)
        willCallRealMethod().given(destinationCell).piece = any<Piece>()

        val enemy = mock<Piece>()
        ReflectionTestUtils.setField(enemy, "color", Color.WHITE)

        val targetCell = mock<Cell>()
        given(targetCell.piece).willReturn(enemy)
        willCallRealMethod().given(targetCell).piece = anyOrNull()

        given(cell.between(eq(destinationCell))).willReturn(targetCell)
        given(cell.diff(eq(destinationCell))).willReturn(2)
        willCallRealMethod().given(cell).piece = anyOrNull()

        ReflectionTestUtils.setField(underTest, "color", Color.BLACK)
        underTest.isCanEat = true

        // when
        underTest.eat(destinationCell)

        // then
        val pieceInSourceCell = ReflectionTestUtils.getField(cell, "piece")
        assertThat(pieceInSourceCell).isNull()

        val pieceInTargetCell = ReflectionTestUtils.getField(targetCell, "piece")
        assertThat(pieceInTargetCell).isNull()

        val pieceInDestinationCell = ReflectionTestUtils.getField(destinationCell, "piece")
        assertThat(pieceInDestinationCell).isNotNull()
        assertThat(pieceInDestinationCell).isInstanceOf(King::class.java)
        assertThat((pieceInDestinationCell as King).color).isEqualTo(Color.BLACK)
    }

    @Test
    fun `creating test`() {
        // when
        val man = Man(Color.BLACK)

        // then
        assertThat(man.color).isEqualTo(Color.BLACK)
    }

    @Test
    fun `isMan test`() {
        // when
        val actual = underTest.isMan()

        // then
        assertThat(actual).isTrue()
    }

    @Test
    fun `isKing test`() {
        // when
        val actual = underTest.isKing()

        // then
        assertThat(actual).isFalse()
    }
}